presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java

```diff
@@ -100,11 +104,15 @@ public PlanNode planTableScan(TableScanNode node, Expression predicate)
                 stripDeterministicConjuncts(predicate),
                 DomainTranslator.toPredicate(layout.getUnenforcedConstraint().transform(assignments::get)));
 
+        PlanNode result = rewrittenTableScan;
         if (!BooleanLiteral.TRUE_LITERAL.equals(resultingPredicate)) {
-            return new FilterNode(idAllocator.getNextId(), result, resultingPredicate);
+            result = new FilterNode(idAllocator.getNextId(), rewrittenTableScan, resultingPredicate);
         }
 
-        return result;
+        if (result instanceof FilterNode && !planChanged(node, rewrittenTableScan, predicate, resultingPredicate)) {
+            return Optional.empty();
+        }
+        return Optional.of(result);
```

Comment:
You don't check if plan has changed if there are no filters, but just table scan